File: src/java/Eratosthenes.java
import java.util.ArrayList;
import java.util.List;

public class Main {
    public static List<Integer> erastothenes_sieve(int upper_limit){
        List<Integer> primes = new ArrayList<Integer>();
        primes.add(2);
        primes.add(3);

        boolean[] bool_array = new boolean[upper_limit];

        int i = 5;
        int adder = 0;

        while(i < upper_limit){
            if(!bool_array[i]) {

                primes.add(i);

                int multiple = i * 3;
                int inc = i * 2;
                while (multiple < upper_limit) {
                    bool_array[multiple] = true;
                    multiple += inc;
                }
            }
            i += 2 + adder;
            adder = (adder - 2) * -1;
        }

        return primes;
    }

    public static void main(String[] args) {
        int len = 1000000;
        int iterations = 1000;
        System.out.println("Starting");
        long time = System.currentTimeMillis();
        for (int i = 0; i < iterations; i++){
            var primes = erastothenes_sieve(len);
        }
        long timePassed = System.currentTimeMillis() - time;
        System.out.println("Finished in " + timePassed / (double) iterations + " milliseconds");
    }
}

File: src/python/eratosthenes.py
import time
# from functools import cache
# from numba import njit

# the performance bootst using this external lib is impressive
# @njit
# @cache
def erasthenos(upper_bound):
    bool_array = [True] * upper_bound
    bool_array[0] = False

    primes = [2, 3]

    i = 4
    adder = 0
    while i < upper_bound:
        if bool_array[i]:
            prime = i + 1
            primes.append(prime)
            multiples = prime * 3

            while multiples < upper_bound:
                bool_array[multiples-1] = False
                multiples += 2 * prime

        i += 2 + adder
        adder = abs(adder - 2)

    return primes


if __name__ == '__main__':
    start_time = time.time()
    iterations = 1000
    length = 1000000
    for i in range(iterations):
        erasthenos(length)
    time_passed = time.time() - start_time
    print(f"finished in {time_passed / iterations * 1000} in milliseconds")